src/gen_rules.ml

```diff
@@ -830,117 +854,106 @@ Add it to your jbuild file to remove this warning.
 
   (* META files that must be installed. Either because there is an explicit or user
      generated one, or because *)
-  let packages_with_explicit_or_user_generated_meta =
+  let init_meta () =
     String_map.values (SC.packages sctx)
-    |> List.filter_map ~f:(fun (pkg : Package.t) ->
+    |> List.iter ~f:(fun (pkg : Package.t) ->
       let path = Path.append ctx.build_dir pkg.path in
-      let meta_fn = "META." ^ pkg.name in
-      let meta_templ_fn = meta_fn ^ ".template" in
-
-      let files =
-        SC.sources_and_targets_known_so_far sctx ~src_path:pkg.path
-      in
-      let has_meta, has_meta_tmpl =
-        (String_set.mem meta_fn files,
-         String_set.mem meta_templ_fn files)
-      in
-
-      let meta_fn =
-        if has_meta then
-          meta_fn ^ ".from-jbuilder"
-        else
-          meta_fn
-      in
-      let meta_path = Path.relative path meta_fn in
-
-      let version =
-        let get =
-          match pkg.version_from_opam_file with
-          | Some s -> Build.return (Some s)
-          | None ->
-            let rec loop = function
-              | [] -> Build.return None
-              | candidate :: rest ->
-                let p = Path.relative path candidate in
-                Build.if_file_exists p
-                  ~then_:(Build.lines_of p
-                          >>^ function
-                          | ver :: _ -> Some ver
-                          | _ -> Some "")
-                  ~else_:(loop rest)
-            in
-            loop
-              [ pkg.name ^ ".version"
-              ; "version"
-              ; "VERSION"
-              ]
+      SC.on_load_dir sctx ~dir:path ~f:(fun () ->
```

Comment:
Why do we need to guard the META generation behind on_load_dir?

Reply:
just to make it lazy